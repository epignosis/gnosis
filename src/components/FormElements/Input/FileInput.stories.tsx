import React from "react";
import { Story } from "@storybook/react";
import FileInputComponent, { FileInputProps } from "./FileInput";

export default {
  title: "components/Form Elements/File Input",
  component: FileInputComponent,
  argTypes: {
    type: {
      control: false,
    },
  },
};

const Template: Story<FileInputProps> = (args) => <FileInputComponent {...args} />;

export const Default = Template.bind({});

Default.args = {
  id: "fileInputStory",
  name: "fileInputStory",
  label: "Input label",
};
