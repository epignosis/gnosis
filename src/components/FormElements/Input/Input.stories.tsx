import React from "react";
import { Story } from "@storybook/react";
import InputComponent, { InputProps } from "./Input";
import { CalendarSVG } from "@icons/core";

export default {
  title: "components/Form Elements/Input",
  component: InputComponent,
  argTypes: {
    size: {
      control: {
        type: "select",
        options: ["md", "lg"],
      },
    },
    status: {
      control: {
        type: "select",
        options: ["valid", "error"],
      },
    },
  },
  args: {
    id: "input",
    size: "md",
    placeholder: "Your LMS username",
    label: "Username",
    inline: false,
    status: "valid",
    className: "inputStory",
    error: "",
  },
  decorators: [
    (Story: Story): JSX.Element => (
      <div style={{ maxWidth: 400 }}>
        <Story />
      </div>
    ),
  ],
};

const Template: Story<InputProps> = (args) => <InputComponent {...args} />;

export const Default = Template.bind({});

export const WithIconBefore = Template.bind({});

WithIconBefore.args = {
  iconBefore: CalendarSVG,
};

export const WithIconAfter = Template.bind({});

WithIconAfter.args = {
  iconAfter: CalendarSVG,
};

export const WithError = Template.bind({});

WithError.args = {
  status: "error",
  error: "This is an inline error",
};
