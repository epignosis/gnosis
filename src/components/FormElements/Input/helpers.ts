export const preventNonNumericalInput = (
  e: React.KeyboardEvent<HTMLInputElement>,
  isNumberType: boolean,
): void => {
  const allowedChars = ["Backspace", "Tab", ".", ",", "ArrowUp", "ArrowDown"];
  const numberPattern = /[0-9]/;

  const isCharacterNumber = numberPattern.test(e.key);
  const isAllowedChar = allowedChars.includes(e.key);

  if (!isCharacterNumber && !isAllowedChar && isNumberType) {
    e.preventDefault();
  }
};
