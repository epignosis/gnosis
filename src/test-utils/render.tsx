import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
