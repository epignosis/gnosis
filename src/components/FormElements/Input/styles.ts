import { css, Theme, SerializedStyles } from "@emotion/react";
import { inputBaseStyles, inputContainerBaseStyles } from "../styles";
import { InputSize } from "./Input";

export const inputContainer = (
  { formElements }: Theme,
  { size }: { size: InputSize },
): SerializedStyles => css`
  ${inputContainerBaseStyles({ block: true })};

  &.error {
    .input-wrapper input {
      border-color: ${formElements.errors.errorBorderColor};
    }
  }

  &.with-prefix-icon {
    .input-wrapper input {
      padding-left: 3.2rem;
    }
  }

  &.with-suffix-icon {
    .input-wrapper input {
      padding-right: 3.2rem;
    }
  }

  .input-wrapper {
    width: 100%;
    position: relative;
    background-color: ${formElements.input.background};
    z-index: 1;

    input {
      ${inputBaseStyles(formElements, { block: true, size })};
      background-color: transparent;
    }

    .prefix-icon,
    .suffix-icon {
      height: 100%;
      position: absolute;
      bottom: 0;
      display: flex;
      align-items: center;
      z-index: -1;
    }

    .prefix-icon {
      margin: 0 0.75rem;
      color: ${formElements.input.iconColor};
    }

    .suffix-icon {
      right: 0;
      padding-right: 0.5rem;
      color: ${formElements.input.iconColor};
    }
  }
`;

export const fileInputContainer = ({ formElements }: Theme): SerializedStyles => css`
  .drop-area {
    position: relative;
    min-height: 5rem;
    border: 2px dashed ${formElements.input.inputBorderColor};
    border-radius: 5px;

    input {
      display: none;
    }
  }
`;
