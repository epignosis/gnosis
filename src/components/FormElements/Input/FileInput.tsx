import React, { FC, InputHTMLAttributes } from "react";
import { Label } from "../../../";
import { fileInputContainer } from "./styles";
import { ExtendableProps } from "types/utils";

export type FileInputProps = ExtendableProps<
  InputHTMLAttributes<HTMLInputElement>,
  {
    id: string;
    name: string;
    label?: string;
    type?: never;
    children?: never;
  }
>;

const FileInput: FC<FileInputProps> = ({ id, label, name, ...rest }) => (
  <div css={fileInputContainer}>
    <label htmlFor={id}>
      <Label id={`${id}-${name}`}>{label}</Label>
      <div className="drop-area">
        <input id={id} type="file" name={name} {...rest} aria-describedby={`${id}-${name}`} />
      </div>
    </label>
  </div>
);

export default FileInput;
