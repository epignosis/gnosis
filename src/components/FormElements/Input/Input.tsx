import React, { forwardRef, Ref, ForwardRefRenderFunction } from "react";
import classNames from "classnames";
import { SerializedStyles } from "@emotion/react";
import Label from "../Label/Label";
import { inputContainer } from "./styles";
import { ExtendableProps, IconType } from "types/common";

export type InputSize = "md" | "lg";

export type InputProps = ExtendableProps<
  React.InputHTMLAttributes<HTMLInputElement>,
  {
    id: string;
    status?: "valid" | "error";
    size?: InputSize;
    iconBefore?: IconType;
    iconAfter?: IconType;
    ref?: Ref<HTMLInputElement>;
    label?: string;
    inline?: boolean;
    containerAttrs?: React.HTMLAttributes<HTMLDivElement>;
  }
>;

const Input: ForwardRefRenderFunction<HTMLInputElement, InputProps> = (
  {
    size = "md",
    status = "valid",
    iconBefore,
    iconAfter,
    label,
    inline = false,
    id,
    containerAttrs,
    ...rest
  },
  forwardedRef,
) => {
  const IconBefore = iconBefore;
  const IconAfter = iconAfter;
  const hasLabel = Boolean(label);
  const containerClasses = classNames({
    valid: status === "valid",
    error: status === "error",
    "with-prefix-icon": Boolean(iconBefore),
    "with-suffix-icon": Boolean(iconAfter),
    inline: hasLabel && inline,
  });

  return (
    <div
      css={(theme): SerializedStyles => inputContainer(theme, { size })}
      className={containerClasses}
      {...containerAttrs}
    >
      {hasLabel && <Label htmlFor={id}>{label}</Label>}
      <div className="input-wrapper">
        {IconBefore && (
          <span className="prefix-icon" data-testid="input-icon-before">
            <IconBefore height={32} />
          </span>
        )}
        <input ref={forwardedRef} id={id} {...rest} />
        {IconAfter && (
          <span className="suffix-icon" data-testid="input-icon-after">
            <IconAfter height={32} />
          </span>
        )}
      </div>
    </div>
  );
};

export default forwardRef(Input);
