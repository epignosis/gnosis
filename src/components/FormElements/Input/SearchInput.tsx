import React, { FC, useEffect, useState, MouseEvent } from "react";
import { useDebounceFn } from "ahooks";
import { MagnifierSVG } from "../../../icons/";
import { searchInputContainer } from "./styles";
import Input from "./Input";

type SearchInputProps = React.HTMLAttributes<HTMLInputElement> & {
  id: string;
  placeholder: string;
  delayBeforeSearch?: number;
  resetInput?: boolean;
  autoFocus?: boolean;
  disabled?: boolean;
  onInputChanged: (inputValue: string) => void;
};

const SearchInput: FC<SearchInputProps> = ({
  id,
  placeholder,
  delayBeforeSearch = 500,
  resetInput = false,
  autoFocus = false,
  disabled = false,
  onInputChanged,
  ...rest
}) => {
  const [searchValue, setSearchValue] = useState("");
  const { run } = useDebounceFn(
    (value) => {
      onInputChanged(value);
    },
    { wait: delayBeforeSearch },
  );

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    const value = e.target.value;
    setSearchValue(value);
    run(value);
  };

  useEffect(() => {
    if (resetInput) clearSearch();
  }, [resetInput]);

  const clearSearch = (e?: MouseEvent<HTMLDivElement>): void => {
    if (e) {
      e.stopPropagation();
    }
    onInputChanged("");
    setSearchValue("");
  };

  return (
    <div css={searchInputContainer} {...rest}>
      <Input
        id={id}
        placeholder={placeholder}
        value={searchValue}
        onChange={handleInputChange}
        autoComplete="off"
        iconAfter={MagnifierSVG}
        showVerticalLine={false}
        isClearable={true}
        autoFocus={autoFocus}
        onClear={(e?: MouseEvent<HTMLDivElement>): void => clearSearch(e)}
        disabled={disabled}
      />
    </div>
  );
};

export default SearchInput;
