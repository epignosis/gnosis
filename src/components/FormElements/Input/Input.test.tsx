import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Input from "./Input";
import { screen, render } from "@test-utils/render";
import { CalendarSVG } from "@icons/core";

describe("<Input />", () => {
  it("renders correctly", () => {
    const mockFn = jest.fn();
    const placeholder = faker.internet.exampleEmail();
    const labelTxt = faker.lorem.word();
    render(
      <Input
        id="test-input"
        name="test-input"
        label={labelTxt}
        placeholder={placeholder}
        value="test"
        onChange={mockFn}
      />,
    );
    const label = screen.getByText(labelTxt);
    const input = screen.getByLabelText(labelTxt);

    expect(label).toHaveTextContent(labelTxt);
    expect(input).toHaveAttribute("name", "test-input");
    expect(input).toHaveAttribute("placeholder", placeholder);
    expect(input).toHaveValue("test");
  });

  it("changes the input's value", () => {
    const labelTxt = faker.lorem.word();
    render(<Input id="test-input" name="test-input" label={labelTxt} />);
    const input = screen.getByLabelText(labelTxt);

    userEvent.type(input, "new value");

    expect(input).toHaveValue("new value");

    userEvent.clear(input);

    expect(input).toHaveValue("");
  });

  it("renders icon before", () => {
    const labelTxt = faker.lorem.word();
    render(<Input id="test-input" name="test-input" label={labelTxt} iconBefore={CalendarSVG} />);

    const iconBefore = screen.getByTestId(/input-icon-before/i);

    expect(iconBefore).toBeInTheDocument();
  });

  it("renders icon after", () => {
    const labelTxt = faker.lorem.word();
    render(<Input id="test-input" name="test-input" label={labelTxt} iconAfter={CalendarSVG} />);

    const iconAfter = screen.getByTestId(/input-icon-after/i);

    expect(iconAfter).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Input id="test-input" name="test-input" label="Test label" iconAfter={CalendarSVG} />,
    );

    expect(container).toMatchSnapshot();
  });
});
