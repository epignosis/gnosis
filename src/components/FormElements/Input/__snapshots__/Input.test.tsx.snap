// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> matches snapshot 1`] = `
<div>
  <div
    class="container-classname css-1it7k89-inputContainerBaseStyles-inputContainerBaseStyles-inputBaseStyles-inputBaseStyles-inputContainer-inputContainer"
    id="container-id"
  >
    <label
      class="css-xs6o3g-label-label"
      for="my-input-id"
    >
      Test label
    </label>
    <div
      class="input-wrapper"
    >
      <input
        class="input"
        id="my-input-id"
        name="test-input"
      />
      <span
        class="suffix-icon"
        data-testid="input-icon-after"
      >
        <svg-file-stub
          height="32"
        />
      </span>
    </div>
  </div>
</div>
`;
